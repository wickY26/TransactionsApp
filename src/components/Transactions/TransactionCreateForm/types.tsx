import { Account } from '../../../containers/Accounts/types';
import { TransactionBody } from '../../../containers/Transactions/types';

export interface TransactionCreateFormProps {
  onCreateTransaction: (data: TransactionBody) => void;
  accounts: Account[];
}
