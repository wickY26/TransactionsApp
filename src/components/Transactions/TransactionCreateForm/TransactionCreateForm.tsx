import React, { FC, FormEvent, ChangeEvent, useState } from 'react';
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import SaveIcon from '@material-ui/icons/Save';
import Autocomplete from '@material-ui/lab/Autocomplete';

import { TransactionCreateFormProps } from './types';
import { Account } from '../../../containers/Accounts/types';

const TransactionCreateForm: FC<TransactionCreateFormProps> = ({ onCreateTransaction, accounts }) => {
  const [from, setFrom] = useState<Account | null>(null);
  const [to, setTo] = useState<Account | null>(null);
  const [amount, setAmount] = useState<string>('');
  const [description, setDescription] = useState<string>('');

  const handleSubmit = (evt: FormEvent<HTMLFormElement>) => {
      evt.preventDefault();
      onCreateTransaction({ 
        from: from?.id!,
        to: to?.id!, 
        amount: parseInt(amount), 
        description,
      });
      setFrom(null);
      setTo(null);
      setAmount('');
      setDescription('');
  }

  const getOptions = (accounts: Account[], selectedId: number | undefined) => accounts.filter(({ id }) => id !== selectedId)

  return (
    <form noValidate autoComplete="off" onSubmit={handleSubmit}>
      <Box display="flex" flexDirection="column" m={2}>
        <Autocomplete
          id="from-autocomplete"
          data-testid="from-autocomplete"
          options={getOptions(accounts, to?.id)} 
          getOptionLabel={({ name, surname }: Account) => `${name} ${surname}`}
          autoComplete
          autoHighlight
          fullWidth
          onChange={(evt: ChangeEvent<{}>, value: Account | null) => setFrom(value)}
          value={from}
          renderInput={(params) => <TextField {...params} inputProps={{ ...params.inputProps, "data-testid": "from"  }} margin="normal" label="From" />}
        />
        <Autocomplete
          id="to-autocomplete"
          data-testid="to-autocomplete"
          options={getOptions(accounts, from?.id)}
          getOptionLabel={({ name, surname }: Account) => `${name} ${surname}`}
          autoComplete
          autoHighlight
          fullWidth
          onChange={(evt: ChangeEvent<{}>, value: Account | null) => setTo(value)}
          value={to}
          renderInput={(params) => <TextField {...params} inputProps={{ ...params.inputProps, "data-testid": "to"  }} margin="normal" label="To" />}
        />
        <TextField 
          inputProps={{ "data-testid": "amount" }}
          margin="normal"
          label="Amount" 
          value={amount}
          type="number"
          onChange={e => setAmount(e.target.value)}/>
        <TextField
          inputProps={{ "data-testid": "description" }}
          margin="normal"
          label="Description" 
          value={description}
          onChange={e => setDescription(e.target.value)}/>
        <Button
          disabled={!from || !to || !amount || !description}
          data-testid="submitBtn"
          variant="contained"
          color="primary"
          startIcon={<SaveIcon />}
          type="submit"
        >
          Create
        </Button>
      </Box>
    </form>
  );
}

export default TransactionCreateForm;
