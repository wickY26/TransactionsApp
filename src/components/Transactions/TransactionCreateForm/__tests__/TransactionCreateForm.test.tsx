import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import TransactionCreateForm from '..';

const mockAccount = {name: 'Poyraz', surname: 'Yilmaz', id: 0};
const mockAccount2 = {name: 'Neslihan', surname: 'Yilmaz', id: 1};
const mockAccounts = [mockAccount, mockAccount2];
const mockOnCreateTransaction = jest.fn();

const renderComponent = () => render(
  <TransactionCreateForm onCreateTransaction={mockOnCreateTransaction} accounts={mockAccounts} />
);

describe('TransactionCreateForm', () => {

  it('should render correctly', () => {
    const component = renderComponent();
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should make submit button disabled when name or surname fields empty', () => {
    renderComponent();

    expect(screen.getByTestId('submitBtn')).toHaveAttribute('disabled');
  });

  it('should call onCreateTransaction when form is submitted', () => {
    renderComponent();

    const fromAutocomplete = screen.getByTestId('from-autocomplete');
    const toAutocomplete = screen.getByTestId('to-autocomplete');

    fireEvent.keyDown(fromAutocomplete, { key: 'ArrowDown' });
    fireEvent.keyDown(fromAutocomplete, { key: 'Enter' });

    fireEvent.keyDown(toAutocomplete, { key: 'ArrowDown' });
    fireEvent.keyDown(toAutocomplete, { key: 'Enter' });

    fireEvent.change(screen.getByTestId('amount'), { target: { value: '250' } });
    fireEvent.change(screen.getByTestId('description'), { target: { value: 'Description' } });
    fireEvent.click(screen.getByTestId('submitBtn'));

    expect(mockOnCreateTransaction).toHaveBeenCalled();
    expect(mockOnCreateTransaction).toHaveBeenCalledWith({from:0, to: 1, amount: 250, description: 'Description'});
  });

  it('should reset form after submit', () => {
    renderComponent();
    const fromField = screen.getByTestId('from') as HTMLInputElement;
    const toField = screen.getByTestId('to') as HTMLInputElement;
    const amountField = screen.getByTestId('amount') as HTMLInputElement;
    const descriptionField = screen.getByTestId('description') as HTMLInputElement;
    const fromAutocomplete = screen.getByTestId('from-autocomplete');
    const toAutocomplete = screen.getByTestId('to-autocomplete');

    fireEvent.keyDown(fromAutocomplete, { key: 'ArrowDown' });
    fireEvent.keyDown(fromAutocomplete, { key: 'Enter' });

    fireEvent.keyDown(toAutocomplete, { key: 'ArrowDown' });
    fireEvent.keyDown(toAutocomplete, { key: 'Enter' });

    fireEvent.change(screen.getByTestId('amount'), { target: { value: '250' } });
    fireEvent.change(screen.getByTestId('description'), { target: { value: 'Description' } });
    fireEvent.click(screen.getByTestId('submitBtn'));

    expect(fromField.value).toEqual('');
    expect(toField.value).toEqual('');
    expect(amountField.value).toEqual('');
    expect(descriptionField.value).toEqual('');
  });

});
