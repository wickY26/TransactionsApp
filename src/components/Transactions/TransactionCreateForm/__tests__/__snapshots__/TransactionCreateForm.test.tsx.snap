// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionCreateForm should render correctly 1`] = `
<DocumentFragment>
  <form
    autocomplete="off"
    novalidate=""
  >
    <div
      class="MuiBox-root MuiBox-root-1"
    >
      <div
        aria-expanded="false"
        class="MuiAutocomplete-root MuiAutocomplete-fullWidth MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
        data-testid="from-autocomplete"
        role="combobox"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink="false"
            for="from-autocomplete"
            id="from-autocomplete-label"
          >
            From
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
          >
            <input
              aria-autocomplete="both"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd"
              data-testid="from"
              id="from-autocomplete"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Clear"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                tabindex="-1"
                title="Clear"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-expanded="false"
        class="MuiAutocomplete-root MuiAutocomplete-fullWidth MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
        data-testid="to-autocomplete"
        role="combobox"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
            data-shrink="false"
            for="to-autocomplete"
            id="to-autocomplete-label"
          >
            To
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
          >
            <input
              aria-autocomplete="both"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd"
              data-testid="to"
              id="to-autocomplete"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Clear"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                tabindex="-1"
                title="Clear"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
        >
          Amount
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            data-testid="amount"
            type="number"
            value=""
          />
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
        >
          Description
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            data-testid="description"
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
        data-testid="submitBtn"
        disabled=""
        tabindex="-1"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
              />
            </svg>
          </span>
          Create
        </span>
      </button>
    </div>
  </form>
</DocumentFragment>
`;
